### DISCUSSION TITLE
Self-Hosted Runner: "once" and "timeout" options

### DISCUSSION THREAD

**Discussion by elliot-nelson** [community/community#28628] [Actions]
Today, the GitHub Action Runner agent only supports two modes: either you start with --ephemeral, which runs exactly one job and then removes your agent, or it runs forever picking up jobs until it is killed or removed.

I would very much like the --once option that the Azure DevOps agent supports: a non-ephemeral agent that you will run multiple jobs on, but, you can take a second to perform custom cleanup actions in between accepting jobs.

In addition to the above I would very much like to have an option that allows me to say "if you have waited for a job for X minutes without getting one, go ahead and exit with a specific error code". This then gives my custom scripting the ability to make a decision about whether to terminate my agent completely, for example.

These two options together would significantly increase the ability to create custom self-hosted auto-scaling solutions that run many jobs on a single agent, and to do so in a way that allows agent cleanup without race conditions.

**Comment by e10harvey** [comment]
We would like these features too.

**Reply by e10harvey** [reply]
For a workaround, see https://github.com/orgs/community/discussions/58215#discussioncomment-9228837.
